workflow/metrics/metrics.go

```diff
@@ -111,19 +114,23 @@ func (m *Metrics) WorkflowAdded(key string, phase v1alpha1.NodePhase) {
 }
 
 func (m *Metrics) WorkflowUpdated(key string, fromPhase, toPhase v1alpha1.NodePhase) {
-	m.workflowsMutex.Lock()
-	hasKey := m.workflows[key]
-	m.workflowsMutex.Unlock()
-	if fromPhase == toPhase || !hasKey {
+	m.mutex.Lock()
+	defer m.mutex.Unlock()
+
+	if fromPhase == toPhase || !m.workflows[key] {
 		return
 	}
-	m.WorkflowDeleted(key, fromPhase)
-	m.WorkflowAdded(key, toPhase)
+	if _, ok := m.workflowsByPhase[fromPhase]; ok {
```

Comment:
is this copy-and-paste? maybe make private func which does not lock?

Reply:
Not quite copy-and-paste, there is some logic that would be redundant if we go with the private func approach; mainly, we will remove and readd the key from the map and perform an additional check unnecessarily. I actually think this is simpler and easier to understand.

However, I don't feel strongly and would gladly go with the private func approach if you think it's necessary.

Reply:
we should just make sure that if we need to make changes in two places - we are aware of it